File: exercises/ddd_infrastructure_layer/typescript/src/application/services/RejectService.ts
import { ApplicationId } from '../../domain/application/ApplicationId';
import type { ApplicationRepository } from '../../domain/application/ApplicationRepository';

export interface RejectCommand {
  applicationId: string; // UUIDv4
}

export class RejectService {
  private repo: ApplicationRepository;

  constructor(repo: ApplicationRepository) {
    this.repo = repo;
  }

  async execute(cmd: RejectCommand) {
    const id = new ApplicationId(cmd.applicationId);
    const application = await this.repo.getById(id);
    if (!application) {
      throw new Error(`Application not found for id ${cmd.applicationId}`);
    }

    try {
      application.reject();
    } catch (e) {
      throw new Error(`Failed to reject application: ${(e as Error).message}`);
    }

    await this.repo.save(application);
  }
}

File: exercises/ddd_infrastructure_layer/java/src/main/java/application/services/RejectService.java
package application.services;

import java.util.UUID;

import com.google.inject.Inject;

import domain.application.ApplicationId;
import domain.application.ApplicationRepository;

public final class RejectService {
    private final ApplicationRepository repository;

    @Inject
    public RejectService(ApplicationRepository repository) {
        this.repository = repository;
    }

    public void execute(RejectCommand command) throws Exception {
        ApplicationId id = new ApplicationId(command.applicationId());
        var application = repository.getById(id);
        if (application == null) {
            throw new IllegalStateException("Application not found for id " + command.applicationId());
        }

        try {
            application.reject();
        } catch (Exception ex) {
            throw new IllegalStateException("Failed to reject application: " + ex.getMessage(), ex);
        }

        repository.save(application);
    }

    public record RejectCommand(UUID applicationId) {
    }
}
